data/src/test/java/org/apache/iceberg/data/avro/TestReadDefaultValues.java

```diff
@@ -0,0 +1,198 @@
+
+  private static final Object[][] typesWithDefaults =
+      new Object[][] {
+        {Types.BooleanType.get(), "true"},
+        {Types.IntegerType.get(), "1"},
+        {Types.LongType.get(), "9999999"},
+        {Types.FloatType.get(), "1.23"},
+        {Types.DoubleType.get(), "123.456"},
+        {Types.DateType.get(), "\"2007-12-03\""},
+        {Types.TimeType.get(), "\"10:15:30\""},
+        {Types.TimestampType.withoutZone(), "\"2007-12-03T10:15:30\""},
+        {Types.TimestampType.withZone(), "\"2007-12-03T10:15:30+00:00\""},
+        {Types.StringType.get(), "\"foo\""},
+        {Types.UUIDType.get(), "\"eb26bdb1-a1d8-4aa6-990e-da940875492c\""},
+        {Types.FixedType.ofLength(2), "\"111f\""},
+        {Types.BinaryType.get(), "\"0000ff\""},
+        {Types.DecimalType.of(9, 4), "\"123.4500\""},
+        {Types.DecimalType.of(9, 0), "\"2\""},
+        // Avro doesn't support negative scale
+        // {Types.DecimalType.of(9, -20), "\"2E+20\""},
+        {Types.ListType.ofOptional(1, Types.IntegerType.get()), "[1, 2, 3]"},
+        {
+          Types.MapType.ofOptional(2, 3, Types.IntegerType.get(), Types.StringType.get()),
+          "{\"keys\": [1, 2], \"values\": [\"foo\", \"bar\"]}"
+        },
+        {
+          Types.StructType.of(
+              required(4, "f1", Types.IntegerType.get()),
+              optional(5, "f2", Types.StringType.get())),
+          "{\"4\": 1, \"5\": \"bar\"}"
+        },
+        // deeply nested complex types
+        {
+          Types.ListType.ofOptional(
+              6,
+              Types.StructType.of(
+                  required(7, "f1", Types.IntegerType.get()),
+                  optional(8, "f2", Types.StringType.get()))),
+          "[{\"7\": 1, \"8\": \"bar\"}, {\"7\": 2, \"8\": " + "\"foo\"}]"
+        },
+        {
+          Types.MapType.ofOptional(
+              9,
+              10,
+              Types.IntegerType.get(),
+              Types.StructType.of(
+                  required(11, "f1", Types.IntegerType.get()),
+                  optional(12, "f2", Types.StringType.get()))),
+          "{\"keys\": [1, 2], \"values\": [{\"11\": 1, \"12\": \"bar\"}, {\"11\": 2, \"12\": \"foo\"}]}"
+        },
+        {
+          Types.StructType.of(
+              required(
+                  13,
+                  "f1",
+                  Types.StructType.of(
+                      optional(14, "ff1", Types.IntegerType.get()),
+                      optional(15, "ff2", Types.StringType.get()))),
+              optional(
+                  16,
+                  "f2",
+                  Types.StructType.of(
+                      optional(17, "ff1", Types.StringType.get()),
+                      optional(18, "ff2", Types.IntegerType.get())))),
+          "{\"13\": {\"14\": 1, \"15\": \"bar\"}, \"16\": {\"17\": \"bar\", \"18\": 1}}"
+        },
+      };
+
+  @Test
+  public void writeAndValidate() throws IOException {
```

Comment:
I think we also want to write a file with explicit `null` values to validate that if the column is present we won't insert the initial default.

Reply:
Added one more test case to test this.